Userland/Libraries/LibWeb/MimeSniff/Resource.cpp

```diff
@@ -0,0 +1,628 @@
+        if (pattern_matched) {
+            if (auto maybe_type = TRY(MimeType::parse(row.mime_type())); maybe_type.has_value())
+                return maybe_type.release_value();
+        }
+    }
+
+    // 3. Let matchedType be the result of executing the image type pattern matching algorithm given resource’s resource header.
+    auto matched_type = TRY(match_an_image_type_pattern(resource.resource_header()));
+
+    // 4. If matchedType is not undefined, return matchedType.
+    if (matched_type.has_value())
+        return matched_type.release_value();
+
+    // 5. Set matchedType to the result of executing the audio or video type pattern matching algorithm given resource’s resource header.
+    matched_type = TRY(match_an_audio_or_video_type_pattern(resource.resource_header()));
+
+    // 6. If matchedType is not undefined, return matchedType.
+    if (matched_type.has_value())
+        return matched_type.release_value();
+
+    // 7. Set matchedType to the result of executing the archive type pattern matching algorithm given resource’s resource header.
+    matched_type = TRY(match_an_archive_type_pattern(resource.resource_header()));
+
+    // 8. If matchedType is not undefined, return matchedType.
+    if (matched_type.has_value())
+        return matched_type.release_value();
+
+    // 9. If resource’s resource header contains no binary data bytes, return "text/plain".
+    if (!any_of(resource.resource_header(), is_binary_data_byte))
+        return MimeType::create("text"_string, "plain"_string);
+
+    // 10. Return "application/octet-stream".
+    return MimeType::create("application"_string, "octet-stream"_string);
+}
+
+}
+
+namespace Web::MimeSniff {
+
+ErrorOr<Resource> Resource::create(ReadonlyBytes data, SniffingContext sniffing_context, bool no_sniff)
+{
+    // NOTE: Non-standard but for cases where pattern matching fails, let's fall back to the safest MIME type.
+    auto default_computed_mime_type = TRY(MimeType::create("application"_string, "octet-stream"_string));
+    auto resource = Resource { data, no_sniff, move(default_computed_mime_type) };
+
+    TRY(resource.supplied_mime_type_detection_algorithm(data));
+    TRY(resource.context_specific_sniffing_algorithm(sniffing_context));
+
+    return resource;
+}
+
+Resource::Resource(ReadonlyBytes data, bool no_sniff, MimeType&& default_computed_mime_type)
+    : m_no_sniff(no_sniff)
+    , m_computed_mime_type(move(default_computed_mime_type))
+{
+    read_the_resource_header(data);
+}
+
+Resource::Resource(Resource&&) = default;
+Resource& Resource::operator=(Resource&&) = default;
+
+Resource::~Resource() = default;
+
+// https://mimesniff.spec.whatwg.org/#supplied-mime-type-detection-algorithm
+ErrorOr<void> Resource::supplied_mime_type_detection_algorithm(ReadonlyBytes)
+{
+    // 1. Let supplied-type be null.
+    String supplied_type;
+
+    // FIXME: 2. If the resource is retrieved via HTTP, execute the following steps:
```

Comment:
`Resource` itself does not have enough information IMO for determining the supplied type, since we don't have the information about the HTTP headers.

Reply:
From our Discord discussion, I may do something like:
```cpp
Resource::create(ReadOnlyBytes data,
                 SniffingContext sniffing_context = SniffingContext::None,
                  StringView scheme,
                  Optional<MimeType> supplied_type = {},
                  bool no_sniff = false);
```

Or because it seems we only need to consider the HTTP scheme as a special case, I can remove the scheme parameter and just assign `m_supplied_mime_type` to `supplied_type`:
```cpp
Resource::create(ReadOnlyBytes data,
                 SniffingContext sniffing_context = SniffingContext::None,
                  Optional<MimeType> supplied_type = {},
                  bool no_sniff = false);
```

With this change though I would have to modify the spec steps to fit with what we have here

Reply:
To my mind a resource should rather be created and the algorithms should be methods on the resource, or the algorithms should take a resource as an argument.  Right now this created method becomes awkward trying to fit all of these use cases.

Another option would be to use function overloading and create multiple versions of `create` to make the caller's life a bit easier. Anyways, I'm not experienced enough with Serenity or C++ to give a definitive answer, so perhaps you could modify the code to make the image sniffing use case possible. Then we could submit both of our PRs for review and get feedback from more experienced people.

Reply:
Keep in mind that I am trying to follow the spec steps closely. Your approaches would allow us to move away some of `create()`'s parameters to make it less of a burden on the callers.

However, due to how the spec seems to insist that we must keep a [computed MIME type field around](https://mimesniff.spec.whatwg.org/#computed-mime-type), when we create a `Resource` with maybe just the data and no_sniff parameters, the caller would be able to use the `computed_mime_type()` getter. This would return an "application/octet-stream" at any point before we utilize the approaches you mentioned.

As much as I don't like my current approach as it hurts readability, I'll make the necessary changes to get the supplied type detection algorithm working so you can use it. I'll keep this discussion open so reviewers can try to find better approaches than this.

Reply:
thanks, @kemzeb , my changes will be pretty trivial though :)

Reply:
Alright, pushed my changes. I'll open my PR and update my PR description with questions I think should be addressed

Reply:
One other approach I can think of right now that would make this API more readable is to separate the `Resource` creation logic and the sniffing logic that gives us the computed MIME type.

So I imagine the caller would first to `Resource::create(...)` passing the data (still need to figure out how we resolve the reading the resource header problem), the scheme, and the supplied type. The creation logic is just there so that we can extract the resource header and run the supplied MIME type detection algorithm. We do not expose any getters in `Resource`.

We then introduce a helper static method `Resource:::sniff()` where we pass the newly-created resource, the `SniffingContext` if needed, and the `no_sniff` flag. This helper would run the sniffing algorithms and return the computed MIME type. To me this would be more readable, but at the cost of the caller having to add two statements instead of one (also, if they try to run the this helper multiple times on the same resource, we can just add a `has_already_been_sniffed` field in `Resource` and used the cached `m_computed_mime_type`).

Something like this:
```cpp
...
auto resource = TRY(MimeSniff::Resource::create(data, "http", extracted_mime_type);
// Resource doesn't expose a getter to m_computed_mime_type, so the caller can't use its default value here
auto computed_mime_type = MimeSniff::Resource::sniff(resource, MimeSniff::Resource::SniffingContext::Image, true);
// sniff() will return a MimeType value
...
```